Fix GNU/Hurd build by avoiding PATH_MAX usage

sthibaul opened the pull request:
#### 598f74dba3e992cdc11d1f3a8f5aa6f53c60fdd7
<pre>
Fix GNU/Hurd build by avoiding PATH_MAX usage
<a href="https://bugs.webkit.org/show_bug.cgi?id=285806">https://bugs.webkit.org/show_bug.cgi?id=285806</a>

Reviewed by Darin Adler.

This is a rework of the patch proposed on
<a href="https://bugs.webkit.org/show_bug.cgi?id=219572">https://bugs.webkit.org/show_bug.cgi?id=219572</a>

GNU/Hurd does not define an arbitrary PATH_MAX limitation.

- Constructed paths can be allocated as appropriate.

- For realpath() calls, since posix 2008 and since long on bsd/linux-like
OS, realpath() can be passed NULL.

* Source/JavaScriptCore/API/tests/testapi.c:
(main): Pass NULL to realpath().
* Source/WTF/wtf/playstation/FileSystemPlayStation.cpp:
(WTF::FileSystemImpl::listDirectorySub): Compute buffer size.
(WTF::FileSystemImpl::realPath): Pass nullptr to realpath().
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::openTemporaryFile): Compute buffer size.
* Source/WebCore/platform/text/hyphen/HyphenationLibHyphen.cpp:
(WebCore::scanDirectoryForDictionaries): Use FileSystem::realPath().

Canonical link: <a href="https://commits.webkit.org/288943@main">https://commits.webkit.org/288943@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e300477a9175937ff6a0ca868387a49d8fd28e3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84804 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4529 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39192 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89943 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35856 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4618 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12504 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65996 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23810 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87849 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3509 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77061 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46270 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3388 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31283 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34930 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77767 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32092 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91319 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83846 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12143 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8862 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74479 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12370 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72875 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73605 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18220 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17975 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16418 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/3591 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12095 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17535 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106239 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11929 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25640 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15423 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13675 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/playstation/FileSystemPlayStation.cpp:
This is not fine as it is using variable-length stack array.

mcatanzaro replied:
(We could catch this automatically by using `-Werror=vla`. But of course that is not something to attempt for the purposes of this pull request.)

darinadler replied:
Yes, this has to be fixed. Maybe the best fix is something like this:
```
newEntry = makeString(path, '/', stringFromFileSystemRepresentation(name));
```

sthibaul replied:
ok, done so

Constellation commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,9 +270,10 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
-        goto end;
+    const char* tmp = temporaryFileDirectory();
+    const char* pre = prefix.utf8().data();
+    char buffer[strlen(tmp) + 1 + strlen(pre) + 6 + 1];
```
Ditto

sthibaul replied:
ok, allocated dynamically

mcatanzaro commented on Source/JavaScriptCore/API/tests/testapi.c:
```diff
@@ -1578,12 +1578,15 @@ int main(int argc, char* argv[])
 #if !OS(WINDOWS)
-    char resolvedPath[PATH_MAX];
-    if (!realpath(argv[0], resolvedPath))
+    char *resolvedPath = realpath(argv[0], NULL);
```
We don't store memory in raw pointers. Save it directly to a smart pointer:

```suggestion
    std::unique_ptr<char, decltype(free)*> resolvedPath(realpath(argv[0], nullptr), free);
```

Then change uses of `resolvedPath` to `resolvedPath.get()` and remove the call to `free()`.

mcatanzaro replied:
Er, sorry, ignore this suggestion. This is a C source file, so it has to be done manually.

mcatanzaro commented on Source/JavaScriptCore/API/tests/testapi.c:
```diff
@@ -1578,12 +1578,15 @@ int main(int argc, char* argv[])
-    if (chdir(newCWD))
-        fprintf(stdout, "Could not chdir to: %s\n", newCWD);
+    else {
+        char *newCWD = dirname(resolvedPath);
```
```suggestion
        char* newCWD = dirname(resolvedPath);
```

sthibaul replied:
https://ews-build.webkit.org/#/builders/38/builds/84388 told me the contrary:

```
ERROR: Source/JavaScriptCore/API/tests/testapi.c:1585:  Declaration has space between * and variable name in char* newCWD  [whitespace/declaration] [3]
```

I was indeed surprised that it's incoherent with other files, but assumed  check-webkit-style knew better, so not sure how to proceed now.

mcatanzaro replied:
Trust the style checker. It's enforcing different style for C than for C++. Confusing but fine.

mcatanzaro commented on Source/JavaScriptCore/API/tests/testapi.c:
```diff
@@ -1578,12 +1578,15 @@ int main(int argc, char* argv[])
+    else {
+        char *newCWD = dirname(resolvedPath);
+        if (chdir(newCWD))
+            fprintf(stdout, "Could not chdir to: %s\n", newCWD);
```
Might as well opportunistically switch to stderr :

```suggestion
            fprintf(stderr, "Could not chdir to: %s\n", newCWD);
```

mcatanzaro replied:
There's only one other use of `fprintf(stdout` in this file and it's two lines up, so might as well fix them both here.

mcatanzaro commented on Source/WTF/wtf/playstation/FileSystemPlayStation.cpp:
```diff
@@ -138,9 +137,10 @@ bool deleteNonEmptyDirectory(const String& path)
-    char resolvedName[PATH_MAX];
-    const char* result = realpath(fsRep.data(), resolvedName);
-    return result ? String::fromUTF8(result) : filePath;
+    char* resolvedName = realpath(fsRep.data(), nullptr);
```
std::unique_ptr<char, decltype(free)*> resolvedPath(realpath(fsRep.data(), nullptr), free);

sthibaul replied:
ok, done so

mcatanzaro commented on Source/WebCore/platform/text/hyphen/HyphenationLibHyphen.cpp:
```diff
@@ -76,11 +76,13 @@ static void scanDirectoryForDictionaries(const char* directoryPath, UncheckedKey
         auto filePath = FileSystem::pathByAppendingComponent(directoryPathString, fileName);
-        char normalizedPath[PATH_MAX];
-        if (!realpath(FileSystem::fileSystemRepresentation(filePath).data(), normalizedPath))
+        char* normalizedPath = realpath(FileSystem::fileSystemRepresentation(filePath).data(), nullptr);
```
Here you should simply use `FileSystem::realPath`.

The JSC API test calls it directly because it doesn't use WebKit internals, and then FilesystemPlayStation.cpp does so to implement `FileSystem::realPath`, but everywhere else can safely use it.

sthibaul replied:
ok, done so

darinadler commented on Source/JavaScriptCore/API/tests/testapi.c:
```diff
@@ -1578,12 +1578,15 @@ int main(int argc, char* argv[])
 #if !OS(WINDOWS)
-    char resolvedPath[PATH_MAX];
-    if (!realpath(argv[0], resolvedPath))
+    char *resolvedPath = realpath(argv[0], NULL);
```
WebKit coding style uses `char*`, not `char *`

sthibaul replied:
As mentioned in another comment above,

https://ews-build.webkit.org/#/builders/38/builds/84388 told me the contrary:

```
ERROR: Source/JavaScriptCore/API/tests/testapi.c:1585:  Declaration has space between * and variable name in char* newCWD  [whitespace/declaration] [3]
```

I was indeed surprised that it's incoherent with other files, but assumed check-webkit-style knew better, so not sure how to proceed now.

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,15 +270,17 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
-    if (snprintf(buffer, PATH_MAX, "%s/%sXXXXXX", temporaryFileDirectory(), prefix.utf8().data()) >= PATH_MAX)
-        goto end;
+    const char* tmp = temporaryFileDirectory();
+    const char* pre = prefix.utf8().data();
```
This is wrong. It returns a pointer to a temporary `CString`'s storage but then that temporary is deallocated, leaving the pointer deallocated. The result of `utf8()` needs to be stored in a local variable.

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,15 +270,17 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
+    const char* tmp = temporaryFileDirectory();
+    const char* pre = prefix.utf8().data();
+    size_t len = strlen(tmp) + 1 + strlen(pre) + 6 + 1;
+    std::unique_ptr<char, decltype(free)*> buffer(static_cast<char*>(malloc(len)), free);
```
This can use `MallocSpan` or any number of other better idioms. Should not call `malloc` and `free` directly like this.

darinadler reviewed (changes requested):
Need to fix the `utf8().get()` mistake.

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,15 +271,17 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
-    if (snprintf(buffer, PATH_MAX, "%s/%sXXXXXX", temporaryFileDirectory(), prefix.utf8().data()) >= PATH_MAX)
-        goto end;
+    const char* tmp = temporaryFileDirectory();
+    CString pre = prefix.utf8();
```
In WebKit coding style we’d typically name this `prefixUTF8`.

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,15 +271,17 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
-    char buffer[PATH_MAX];
-    if (snprintf(buffer, PATH_MAX, "%s/%sXXXXXX", temporaryFileDirectory(), prefix.utf8().data()) >= PATH_MAX)
-        goto end;
+    const char* tmp = temporaryFileDirectory();
```
In WebKit coding style we’d typically name this something more like `directory`.

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,15 +271,17 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
-        goto end;
+    const char* tmp = temporaryFileDirectory();
+    CString pre = prefix.utf8();
+    size_t len = strlen(tmp) + 1 + pre.length() + 6 + 1;
```
In WebKit coding style we’d normally name this something like `length`.

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,15 +271,17 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
+    CString pre = prefix.utf8();
+    size_t len = strlen(tmp) + 1 + pre.length() + 6 + 1;
+    auto buffer = MallocSpan<char>::malloc(len);
+    snprintf(buffer.mutableSpan().data(), len, "%s/%sXXXXXX", tmp, pre.data());
```
I noticed that this will malfunction if the prefix happens to end with an "X", which we could fix by including at least one non-X character before "XXXXXX".

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -270,15 +271,17 @@ std::pair<String, PlatformFileHandle> openTemporaryFile(StringView prefix, Strin
     // This is OK for now since the code using it is built on macOS only.
```
The comment above saying we can’t implement the suffix is incorrect, because we could do it by using `mkostemps` instead of `mkostemp`.

darinadler reviewed (approved):
Fine as is; I did notice a few things we could improve, but I don’t insist on any of it.

sthibaul committed:
Fix GNU/Hurd build by avoiding PATH_MAX usage
https://bugs.webkit.org/show_bug.cgi?id=285806

Reviewed by Darin Adler.

This is a rework of the patch proposed on
https://bugs.webkit.org/show_bug.cgi?id=219572

GNU/Hurd does not define an arbitrary PATH_MAX limitation.

- Constructed paths can be allocated as appropriate.

- For realpath() calls, since posix 2008 and since long on bsd/linux-like
OS, realpath() can be passed NULL.

* Source/JavaScriptCore/API/tests/testapi.c:
(main): Pass NULL to realpath().
* Source/WTF/wtf/playstation/FileSystemPlayStation.cpp:
(WTF::FileSystemImpl::listDirectorySub): Compute buffer size.
(WTF::FileSystemImpl::realPath): Pass nullptr to realpath().
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::openTemporaryFile): Compute buffer size.
* Source/WebCore/platform/text/hyphen/HyphenationLibHyphen.cpp:
(WebCore::scanDirectoryForDictionaries): Use FileSystem::realPath().

Canonical link: https://commits.webkit.org/288943@main

rkirsling commented on Source/WTF/wtf/playstation/FileSystemPlayStation.cpp:
```diff
@@ -138,9 +134,8 @@ bool deleteNonEmptyDirectory(const String& path)
-    char resolvedName[PATH_MAX];
-    const char* result = realpath(fsRep.data(), resolvedName);
-    return result ? String::fromUTF8(result) : filePath;
+    std::unique_ptr<char, decltype(free)*> resolvedPath(realpath(fsRep.data(), nullptr), free);
+    return resolvedPath ? String::fromUTF8(resolvedPath) : filePath;
```
I don't know what this code is meaning to do, but `String::fromUTF8` is not defined for whatever `std::unique_ptr<char, decltype(free)*>` is.

rkirsling replied:
Ahh, I see an example of this pattern in glib code. Looked more complicated but guess it's just missing a `.get()`?

darinadler replied:
I think that's right. The `get()` is missing.
